### ISSUE TITLE
WALL-E graphical issue for Work Day cutscene

### ISSUE THREAD

**Issue by Panderner** [hrydgard/ppsspp#14168]
In Work Day the 4th level for WALL-E there's a graphical artifact during a cutscene

OpenGL:
![Screenshot_2021-02-17-16-10-28-403_org ppsspp ppsspp](https://user-images.githubusercontent.com/37503397/108175008-299b8800-713b-11eb-8c0c-7778ab51be79.jpg)

Vulkan:
![Screenshot_2021-02-17-16-09-51-432_org ppsspp ppsspp](https://user-images.githubusercontent.com/37503397/108175022-2ef8d280-713b-11eb-8187-9c2b9d457e3f.jpg)

GE dump: 
[ULUS10350_0001.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/5994228/ULUS10350_0001.ppdmp.zip)

**Comment by unknownbrackets** [comment]
This definitely seems depth clamp related.  Draw 73/134 uses especially wild depth values, and causes this glitch.  This isn't the only prim that causes problems, but it's the main one.

It should look like OpenGL:
![ULUS10350_#14168_walle_workday_glitch](https://user-images.githubusercontent.com/191233/108305932-f0294380-715f-11eb-9356-426dc4bb5e7a.png)

Which is how Vulkan looks if I force the depth clamp flag off.  Min Z/max Z are full range, so this isn't a clamp skew issue, it must be a case where it shouldn't *actually* clamp.

-[Unknown]

**Comment by unknownbrackets** [comment]
I tried to implement the logic from https://github.com/hrydgard/ppsspp/issues/12058#issuecomment-913225641 in the software renderer, and test in a few scenarios... but this one is not rendering correctly there.  So I suspect there's more to it in some way.  It might be that w is still involved, after all (although I'd tried a test that should've validated that, there's more cases to try...)

This particular draw is 966 verts, which makes it harder to intuit which triangle is being incorrectly drawn... I've been creating tests with specific triangles from dumps manually.  Maybe I need to make some tool to slice and dice frame dumps...

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, I have a feeling that this is one of those issues where we'll come up with a bunch of complicated rules to figure out what shouldn't be drawn, and there's really a smaller number of much simpler rules that could explain it all, but that are very hard to intuit from the data...

**Comment by unknownbrackets** [comment]
Perhaps, but it does seem to have pretty specific rules.  Maybe there's indeed something simpler about it all..

The reason for this was that I'd removed the Z clipping when changing it (I actually did that intentionally so I'd notice if it should be kept), and it indeed should be kept for the negative side:
https://github.com/hrydgard/ppsspp/issues/12058#issuecomment-913937188

Still hoping to get some time to compare back with guardband culling.  For example, when a vertex's depth is clamped, that vertex also becomes immune to guardband culling.  But is this behavior actually driven off the pre-viewport Z?  This might explain some cases where guardband culling is overzealous (when viewport Z scale is 32768 or similar.)

-[Unknown]

**Comment by Panderner** [comment]
This also affects Direct3D 9 and 11.
Here's a screenshot for D3D9:
![Screenshot (14)](https://user-images.githubusercontent.com/37503397/132284997-5deade33-6eb8-4719-87f3-a9c2759af854.png)

**Comment by hrydgard** [comment]
> But is this behavior actually driven off the pre-viewport Z?

That indeed seems likely from your findings! And the immunity is not likely to be actual 100% immunity, but those long thing triangles that usually are in danger of the guardband are produces by large perspective distortion of triangles crossing the view plane beyond the edges of the viewport, and clipping will cause the new triangle corners to get back inside the guardband. But for practical purposes I think we can emulate that with immunity.

**Comment by unknownbrackets** [comment]
Well, that's what I meant by immunity, but indeed maybe there are actual cases where a triangle might get clipped (depending on how the clipper works...) with it still outside the guardband.  Hopefully we don't need to worry about that...

-[Unknown]

**Comment by hrydgard** [comment]
Yes I think there are some theoretical cases since the clipper is still against the (pre-viewport) Z plane, and not against any side planes, but we can indeed probably just ignore that.

**Comment by Panderner** [comment]
Fixed in Direct3D 11 but not Direct3D 9.

**Comment by Gamemulatorer** [comment]
WALL-E can't get the binoculars

PPSSPP

https://user-images.githubusercontent.com/37603562/191388181-cddc4bd9-a042-4421-8ee6-c6423785fb07.mp4

PSP

https://user-images.githubusercontent.com/37603562/191388226-4fd8fc26-c69e-4d39-b5cc-8121998f5fce.mp4

**Comment by Gamemulatorer** [comment]
I cannot reproduce this issue on my Oppo a3s adreno 506 GPU might be fixed by https://github.com/hrydgard/ppsspp/pull/16068 or https://github.com/hrydgard/ppsspp/pull/16043

Is Direct3D 9 can be fixable?

**Comment by unknownbrackets** [comment]
Direct3D 9 should look better if you turn off hardware transform, since it'll apply cull distance for you.  It won't apply clip distance, though, so not all of the artifacts will be fixed.  In theory, we could clip triangles in software transform, but it's not implemented.

It also won't work on any device without clip distance and working geometry shaders, which is mostly older Mali devices and some other devices.

Modern GPUs that support clip and cull distance (including Adreno) should be fine and no longer have this issue for some time now.

-[Unknown]
